stan/math/prim/prob/hmm_marginal_lpdf.hpp

```diff
@@ -0,0 +1,234 @@
+                    col(Gamma_transpose, i + 1));
+    }
+  }
+  check_consistent_size("hmm_marginal_lpdf", "rho", rho, n_states);
+  check_simplex("hmm_marginal_lpdf", "rho", rho);
+
+  using T_partials_return = partials_return_t<T_omega, T_Gamma, T_rho>;
+  operands_and_partials<Eigen::Matrix<T_omega, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_Gamma, Eigen::Dynamic, Eigen::Dynamic>,
+                        Eigen::Matrix<T_rho, Eigen::Dynamic, 1> >
+      ops_partials(log_omegas, Gamma, rho);
+
+  Eigen::MatrixXd alphas(n_states, n_transitions + 1);
+  Eigen::VectorXd alpha_log_norms(n_transitions + 1);
+  Eigen::MatrixXd omegas;
+  Eigen::MatrixXd Gamma_dbl = value_of_rec(Gamma);
+
+  T_partials_return log_marginal_density
+      = hmm_marginal_lpdf(value_of_rec(log_omegas), Gamma_dbl,
+                          value_of_rec(rho), alphas, alpha_log_norms, omegas);
+
+  // Variables required for all three Jacobian-adjoint products.
+  double norm_norm = alpha_log_norms(n_transitions);
+  double unnormed_marginal = alphas.col(n_transitions).sum();
+
+  std::vector<Eigen::VectorXd> kappa(n_transitions);
+  Eigen::VectorXd kappa_log_norms(n_transitions);
+  std::vector<double> grad_corr(n_transitions);
+
+  if (n_transitions > 0) {
+    kappa[n_transitions - 1] = Eigen::VectorXd::Ones(n_states);
+    kappa_log_norms(n_transitions - 1) = 0;
+    grad_corr[n_transitions - 1]
+        = std::exp(alpha_log_norms(n_transitions - 1) - norm_norm);
+  }
+
+  for (int n = n_transitions - 2; n >= 0; --n) {
+    kappa[n] = Gamma_dbl * (omegas.col(n + 2).cwiseProduct(kappa[n + 1]));
+
+    double norm = kappa[n].maxCoeff();
+    kappa[n] /= norm;
+    kappa_log_norms(n) = std::log(norm) + kappa_log_norms(n + 1);
+    grad_corr[n]
+        = std::exp(alpha_log_norms(n) + kappa_log_norms(n) - norm_norm);
+  }
+
+  if (!is_constant_all<T_Gamma>::value) {
+    Eigen::MatrixXd Gamma_jacad(n_states, n_states);
+    Gamma_jacad.setZero();
+
+    if (n_transitions != 0) {
+      for (int n = n_transitions - 1; n >= 0; --n) {
+        Gamma_jacad += (grad_corr[n] * kappa[n].cwiseProduct(omegas.col(n + 1))
+                        * alphas.col(n).transpose())
+                           .transpose();
+      }
+    }
+
+    Gamma_jacad /= unnormed_marginal;
+    ops_partials.edge2_.partials_ = Gamma_jacad;
+  }
+
+  bool sensitivities_for_omega_or_rho
+      = (!is_constant_all<T_omega>::value) || (!is_constant_all<T_rho>::value);
+
+  // boundary terms
+  if (sensitivities_for_omega_or_rho) {
+    Eigen::MatrixXd log_omega_jacad(n_states, n_transitions + 1);
+    log_omega_jacad.setZero();
+
+    if (!is_constant_all<T_omega>::value) {
+      for (int n = n_transitions - 1; n >= 0; --n)
+        log_omega_jacad.col(n + 1)
+            = grad_corr[n]
+              * kappa[n].cwiseProduct(Gamma_dbl.transpose() * alphas.col(n));
+    }
+
+    // Boundary terms
+    // TO DO (charlesm93): find a better solution that the if loop
```

Comment:
This isn't grammatical as a sentence and the Google style is `TODO(name)` so it's easy to auto-parse.